Fix shift behaviour for ComponentContainers

Holzhaus opened the pull request:
For advanced controller mappings, it can be desirable to check the
current shift state of a component container by reading its `isShifted`
attribute. This attribute is defined as follows:

    isShifted: false,
    shift: function () {
        this.forEachComponent(function (component) {
            if (typeof component.shift === 'function') {
                //[...]
                component.shift();
            }
            // Set isShifted for child ComponentContainers forEachComponent is iterating through recursively
            this.isShifted = true;
        });
    },
    unshift: function () {
        this.forEachComponent(function (component) {
            if (typeof component.unshift === 'function') {
                // [...]
                component.unshift();
            }
            // Set isShifted for child ComponentContainers forEachComponent is iterating through recursively
            this.isShifted = false;
        });
    },

Unfortunately, the comment in the source code is wrong, because `isShifted` will
*not* be set recursively. In fact, it will only ever be set if the
`shift` method of a component container is called directly. When
interating thought components recursively, `forEachComponent` will only
call `shift` for the components inside a component container, but not
for the container itself:

    var applyOperationTo = function (obj) {
        if (obj instanceof Component) {
            operation.call(that, obj);
        } else if (recursive && obj instanceof ComponentContainer) {
            obj.forEachComponent(operation);
        } else if (Array.isArray(obj)) {
            obj.forEach(function (element) {
                applyOperationTo(element);
            });
        }
    };

Hence, `isShifted` will not be set recursively.
To fix that, this PR introduces an additional argument to
`forEachComponent` to also apply the function to containers.
Just doing this unconditionally might break existing controller mappings since
containers do not inherit from `Component` and do not implement the same
methods.

Holzhaus commented:
By the way, this would also allow to have component containers that do not pass the shift state to its children by overriding the `shift` and `unshift` methods with no-ops.

uklotzde reviewed (commented):
General thoughts: From my own experience pressing _Shift_ is a global action and it should be recursively applied to all controls unless some explicit barrier prevents this. Wouldn't it make sense to introduce a breaking change instead of preserving the current, unintuitive behavior?

Be-ing commented on res/controllers/midi-components-0.0.js:
```diff
@@ -503,19 +503,25 @@
                 return;
             }
             if (recursive === undefined) { recursive = true; }
+            if (including_containers === undefined) { including_containers = false; }
```
Would there be a harm in making this true by default? Making it true by default would remove the need to explicitly specify the last 2 parameters in the Button.prototype.shift/unshift functions.

Be-ing replied:
> Just doing this unconditionally might break existing controller mappings since
containers do not inherit from Component and do not implement the same
methods.

Yeah, I think you are right. On further thought, this seems hacky to repurpose a function intended to be used on Components to also be used for ComponentContainers... maybe there is another solution?

Be-ing replied:
How about adding an optional parameter to `forEachComponent` that is a function that gets called on ComponentContainers (in addition to the recursive call to `forEachComponent`)? That way, the shift/unshift functions could pass simple functions like:
`function () { this.isShifted = true; }`
This would keep a separation of concerns between the functions that act on Components and those that act on ComponentContainers.

Holzhaus replied:
> How about adding an optional parameter to forEachComponent that is a function that gets called on ComponentContainers (in addition to the recursive call to forEachComponent)?

Also feasible, but IMHO this is just as "hacky" as this solution. Also, this messes up the signature a bit `forEachComponent(func on_component, bool recusive, func on_container)`.

> On further thought, this seems hacky to repurpose a function intended to be used on Components to also be used for ComponentContainers... maybe there is another solution?

Another alternative might be to write a new method `forEachComponentContainer`. That would keep those concern separated, keep backwards compatibility and look cleaner overall. Downside: A few lines of duplicated code.
In `ComponentContainer.shift()`, we could then just set `recursive` to `false` and invoke `this.forEachComponentContainer(function(container) { container.shift(); })`.

Be-ing replied:
> Another alternative might be to write a new method forEachComponentContainer. That would keep those concern separated, keep backwards compatibility and look cleaner overall.

I like this idea.

Holzhaus replied:
Ok, I reverted my original attempt and pushed a new implementation.

Be-ing commented on res/controllers/midi-components-0.0.js:
```diff
@@ -566,7 +572,7 @@
                 // Set isShifted for child ComponentContainers forEachComponent is iterating through recursively
                 this.isShifted = true;
-            });
+            }, false, true);
```
Unfortunately JS does not support named parameters. Any unspecified trailing parameters to a JS function have a value of `undefined`, which can be used to add optional parameters to end of the parameter list for a function. If a value other than the default is needed, the trailing optional parameters all need to be explicitly specified (the order of parameters matters because the parameter names do not).

Holzhaus committed:
Fix shift behaviour for ComponentContainers

For advanced controller mappings, it can be desireable to check the
current shift state of a component container by reading its `isShifted`
attribute. This attribute is defined as follows:

    isShifted: false,
    shift: function () {
        this.forEachComponent(function (component) {
            if (typeof component.shift === 'function') {
                //[...]
                component.shift();
            }
            // Set isShifted for child ComponentContainers forEachComponent is iterating through recursively
            this.isShifted = true;
        });
    },
    unshift: function () {
        this.forEachComponent(function (component) {
            if (typeof component.unshift === 'function') {
                // [...]
                component.unshift();
            }
            // Set isShifted for child ComponentContainers forEachComponent is iterating through recursively
            this.isShifted = false;
        });
    },

Unfortunately, the comment in the source code is wrong, because `isShifted` will
*not* be set recursively. In fact, it will only ever be set if the
`shift` method of a component container is called directly. When
interating thought components recursively, `forEachComponent` will only
call `shift` for the components inside a component container, but not
for the container itself:

    var applyOperationTo = function (obj) {
        if (obj instanceof Component) {
            operation.call(that, obj);
        } else if (recursive && obj instanceof ComponentContainer) {
            obj.forEachComponent(operation);
        } else if (Array.isArray(obj)) {
            obj.forEach(function (element) {
                applyOperationTo(element);
            });
        }
    };

Hence, `isShifted` will not be set recursively.
To fix that, this PR introduces an additional function
`forEachComponentContainer` that will be used by the
`shift()`/`unshift()` methods to propagate the shift state to child
containers recursively.

Be-ing commented:
Can you rebase to remove the reverted commit? Also, I think this should be rebased on the 2.2 branch.

Holzhaus commented:
@Be-ing Done.

Holzhaus commented:
Is anything else blocking the merge?

uklotzde commented:
@Be-ing ping

Be-ing commented:
I tested this with my Allen & Heath Xone K2s and Hercules P32 and did not find any issues. Thanks for working on this.
